### PR TITLE
Injecting Experiment Data

### PR DESCRIPTION
**Pull request by okaycj**
# Summary

We found an issue where experiment trials couldn't be identified.  For example, our exit survey plugin is an extension of the jsPsych's survey plugin.  This plugin will add its type to the experiment data, but it doesn't know it is an "Exit Survey", just a "Survey".  We add a value to the experiment data to identify both Exit and Consent surveys  

## Implementation 

We exploit the "data" attribute on trials to add data to an experiment.  From the experimenter's perspective, the data attribute could be used like this:

```javascript
{ 
  type: ... Some jsPsych experiment type,
  data: { key: "some data for this trial" } 
}
```

We intercept the timeline and check if the type object has the method "chsData" and add its data to the trial before the experiment runs.

### CODE REVIEW

File: packages/lookit-initjspsych/src/types.ts [hunk 1]
```diff
@@ -1,17 +1,17 @@
...
 };
 
-export type JsPsychOptions = {
-  on_data_update?: UserFunc;
-  on_finish?: UserFunc;
+export type Timeline = {
+  type: {
+    chsData?: () => object;
+  };
+  data?: object;
```
Comment (becky-gilbert) [inline packages/lookit-initjspsych/src/types.ts:16]: I was initially really confused by this type, but I think I get it now... Just to confirm:
This is the type for the trial objects inside the array that's passed to `jsPsych.run`? And each trial object must have a `type` (the plugin class name), which may or may not have a `chsData` static method on the plugin class.
These objects may or may not have a `data` property that is defined by the researcher.
And of course they can have a bunch of other properties (parameters) that aren't listed here. 
Is that right?

**Review by becky-gilbert — APPROVED** [review]
Looks great! One question for my own edification 🙂

File: packages/data/src/types.ts [hunk 2]
```diff
@@ -55,10 +53,15 @@ export interface ChildAttrs extends Attributes {
   readonly pk?: number;
 }
 
-export interface PastSessionAttrs extends Attributes {
+export interface JsPsychExpData {
+  trial_index: number;
+  trial_type: string;
+}
```
[the remark this answers is not in the document]
Comment (okaycj) [reply packages/data/src/types.ts:59]: The provided type for the `values` method on a DataCollection object is `any[]`.  I lean away from `any` as a type.  `JsPsychExpData` just replaces the `any`.

File: packages/lookit-initjspsych/src/types.ts (hunk 1 above)
Comment (okaycj) [reply packages/lookit-initjspsych/src/types.ts:16]: You're correct.  The actual jsPsych type for a timeline object isn't exported.  Or at least, I couldn't find where it was. I would rather use the types provided by jsPsych.

**Pull request merged**